src/lib/cert/x509/name_constraint.cpp

```diff
@@ -0,0 +1,269 @@
+#include <sstream>
+
+namespace Botan {
+
+GeneralName::GeneralName(const std::string& v) : GeneralName()
+   {
+   size_t p = v.find(':');
+
+   if(p != std::string::npos)
+      {
+      m_type = v.substr(0,p);
+      m_name = v.substr(p + 1,std::string::npos);
+      }
+   else
+      {
+      throw Invalid_Argument("Failed to decode Name Constraint");
+      }
+   }
+
+void GeneralName::encode_into(class DER_Encoder&) const
+   {
+   throw std::runtime_error("General Name encoding not implemented");
+   }
+
+void GeneralName::decode_from(class BER_Decoder& ber)
+   {
+   BER_Object obj = ber.get_next_object();
+   if((obj.class_tag != CONTEXT_SPECIFIC) &&
+      (obj.class_tag != (CONTEXT_SPECIFIC | CONSTRUCTED)))
+      throw Decoding_Error("Invalid class tag while decoding GeneralName");
+
+   const ASN1_Tag tag = obj.type_tag;
+
+   if(tag == 1 || tag == 2 || tag == 6)
+      {
+      m_name = Charset::transcode(ASN1::to_string(obj),LATIN1_CHARSET,LOCAL_CHARSET);
+
+      if(tag == 1)
+         {
+         m_type = "RFC822";
+         }
+      else if(tag == 2)
+         {
+         m_type = "DNS";
+         }
+      else if(tag == 6)
+         {
+         m_type = "URI";
+         }
+      }
+   else if(tag == 4)
+      {
+      X509_DN dn;
+      std::multimap<std::string, std::string> nam;
+      BER_Decoder dec(obj.value);
+      std::stringstream ss;
+
+      dn.decode_from(dec);
+      ss << dn;
+
+      m_name = ss.str();
+      m_type = "DN";
+      }
+   else if(tag == 7)
+      {
+      if(obj.value.size() == 8)
+         {
+         const std::vector<byte> ip(obj.value.begin(),obj.value.begin() + 4);
+         const std::vector<byte> net(obj.value.begin() + 4,obj.value.end());
+         m_type = "IP";
+         m_name = ipv4_to_string(load_be<u32bit>(ip.data(),0)) + "/" + ipv4_to_string(load_be<u32bit>(net.data(),0));
+         }
+      }
+   else
+      {
+      throw Decoding_Error("Found unknown GeneralName type");
+      }
+   }
+
+GeneralName::MatchResult GeneralName::matches(const X509_Certificate& cert) const
+   {
+   if(type() == "DNS" || type() == "DN" || type() == "IP")
+      {
+      std::vector<std::string> nam;
+      std::function<bool(const GeneralName*,const std::string&)> match_fn;
+
+      if(type() == "DNS")
+         {
+         match_fn = std::mem_fn(&GeneralName::matches_dns);
+         nam = cert.subject_info("DNS");
+
+         if(nam.empty())
+            {
+            nam = cert.subject_info("CN");
+            }
+         }
+      else if(type() == "DN")
+         {
+         std::stringstream ss;
+
+         match_fn = std::mem_fn(&GeneralName::matches_dn);
+         ss << cert.subject_dn();
+
+         nam.push_back(ss.str());
+         }
+      else if(type() == "IP")
+         {
+         match_fn = std::mem_fn(&GeneralName::matches_ip);
+         nam = cert.subject_info("IP");
+         }
+
+      else
+         {
+         throw Unsupported_Argument("Unsupported Name Constraint");
```

Comment:
The first `if` means this throw is never reached, instead `UnknownType` is returned.